proposals/roles.md

```diff
@@ -0,0 +1,665 @@
+- If no candidate set is found in any enclosing namespace declaration or compilation unit, 
+  the result of the lookup is empty.
+
+TODO2 explain static usings
+
+The preceding rules mean that:
+1. extension members available in inner type declarations take precedence over
+extension members available in outer type declarations,
+2. extension members available in inner namespace declarations 
+take precedence over extension members available in outer namespace declarations,
+3. and that extension members declared directly in a namespace take precedence over 
+extension members imported into that same namespace with a using namespace directive.  
+
+The difference between invocation and non-invocation handling is that for invocation scenarios, 
+we can look past a result and continue looking at enclosing namespaces.  
+
+For example, if an "inner" extension has a method `void M(int)` and an "outer" extension
+has a method `void M(string)`, an extension member lookup for `M("hello")` will look over
+the `int` extension and successfully find the `string` extension.  
+On the other hand, if an "inner" extension has an `int` property and an "outer" extension
+has a string property, an assignment of a string to that property will fail, as
+extension member lookup will find the `int` property and stop there.  
+
+For context see [extension method invocation rules](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#11783-extension-method-invocations).
+
+### Element access
+
+https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#117103-indexer-access
+TODO
+
+### Operators
+
+TODO
+User-defined conversion should be allowed, except where it conflicts with a built-in
+conversion (such as with an underlying type).  
+
+### Method group conversions
+
+https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/conversions.md#108-method-group-conversions
+TODO A single method is selected corresponding to a method invocation, 
+but with some tweaks related to normal form and optional parameters.  
+TODO There's also the scenario where a method group contains a single method (lambda improvements).  
+
+## Implementation details
+
+Roles are implemented as ref structs.  
+If the role any instance member, then we emit a ref field (of underlying type)
+into the ref struct and a constructor.  
+Values of role types are left as values of the underlying value type, until a role
+member is accessed. When a role member is accessed, a role instance is created
+with a reference to the underlying value and the member is accessed on that instance.
+
+```
+Role r = default(UnderlyingType); // emitted as a local of type `UnderlyingType`
+r.RoleMember(); // emitted as `new Role(ref r).RoleMember();`
+```
+
+Roles appearing in signatures are emitted as the role's underlying type
+marked with a modopt of the role type.
+
+```
+void M(Role r) // emitted as `void M(modopt(Role) UnderlyingType r)`
```

Comment:
More detail is needed for scenarios where roles are used as nested types--array types, pointed-at types, type arguments, etc.